go/vt/workflow/sleep_workflow.go

```diff
@@ -0,0 +1,236 @@
+func (sw *SleepWorkflow) Action(ctx context.Context, name string) error {
+	sw.mu.Lock()
+	defer sw.mu.Unlock()
+
+	// Change our state.
+	log.Infof("SleepWorkflow.Action(%v) called.", name)
+	paused := sw.data.Paused
+	switch name {
+	case pauseAction:
+		sw.data.Paused = true
+		sw.logger.Infof("Paused")
+	case resumeAction:
+		sw.data.Paused = false
+		sw.logger.Infof("Resumed")
+	default:
+		sw.logger.Errorf("Unknown action %v called", name)
+		return fmt.Errorf("unknown action %v", name)
+	}
+	if paused == sw.data.Paused {
+		// Nothing changed.
+		return nil
+	}
+
+	// UI update and Checkpoint.
+	sw.node.Modify(func() {
+		sw.uiUpdate()
+	})
+	return sw.checkpoint(ctx)
+}
+
+// UI update every second.
+// Needs to be called with the lock / inside sw.node.Modify.
+func (sw *SleepWorkflow) uiUpdate() {
+	sw.node.Progress = 100 * sw.data.Slept / sw.data.Duration
+	sw.node.ProgressMessage = fmt.Sprintf("%v/%v", sw.data.Slept, sw.data.Duration)
+	sw.node.Log = sw.logger.String()
+	if sw.data.Paused {
+		sw.node.Actions[0].State = ActionStateDisabled
+		sw.node.Actions[1].State = ActionStateEnabled
+		sw.node.ProgressMessage += " (paused)"
+	} else {
+		sw.node.Actions[0].State = ActionStateEnabled
+		sw.node.Actions[1].State = ActionStateDisabled
+	}
+}
+
+// checkpoint saves a checkpoint in topo server.
+// Needs to be called with the lock.
+func (sw *SleepWorkflow) checkpoint(ctx context.Context) error {
+	var err error
+	sw.wi.Data, err = json.Marshal(sw.data)
+	if err != nil {
+		return err
+	}
+	err = sw.manager.TopoServer().SaveWorkflow(ctx, sw.wi)
+	if err != nil {
+		sw.logger.Errorf("SaveWorkflow failed: %v", err)
+	} else {
+		sw.logger.Infof("SaveWorkflow successful")
+	}
+	return err
+}
+
+// SleepWorkflowFactory is the factory to register the Sleep workflows.
+type SleepWorkflowFactory struct{}
+
+// Init is part of the workflow.Factory interface.
+func (f *SleepWorkflowFactory) Init(w *workflowpb.Workflow, args []string) error {
+	// Parse the flags.
+	subFlags := flag.NewFlagSet(sleepFactoryName, flag.ContinueOnError)
+	duration := subFlags.Int("duration", 30, "How long to sleep")
+	if err := subFlags.Parse(args); err != nil {
+		return err
+	}
+	if *duration <= 0 {
+		return fmt.Errorf("duration cannot be a negative number")
+	}
+
+	// Build the Workflow object.
+	w.Name = fmt.Sprintf("Sleep(%v seconds)", *duration)
+	data := &SleepWorkflowData{
+		Duration: *duration,
+	}
+	var err error
+	w.Data, err = json.Marshal(data)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Instantiate is part of the workflow.Factory interface.
+func (f *SleepWorkflowFactory) Instantiate(w *workflowpb.Workflow) (Workflow, error) {
+	data := &SleepWorkflowData{}
+	if err := json.Unmarshal(w.Data, data); err != nil {
+		return nil, err
+	}
+	return &SleepWorkflow{
+		data:   data,
+		logger: logutil.NewMemoryLogger(),
+	}, nil
+}
+
+// Compile time interface check.
+var _ Factory = (*SleepWorkflowFactory)(nil)
```

Comment:
You're saying with this, the 'unused' tool will not report this implementation ever as being unused? I don't mind moving this check to a unit test to fix this. I like the compile version however, it's easier to find and the error message is nice and explicit when there is a mismatch.